src/Compilers/CSharp/Test/Semantic/SourceGeneration/StateTableTests.cs

```diff
@@ -0,0 +1,260 @@
+            AssertTableEntries(table, expected);
+        }
+
+        [Fact]
+        public void Node_Table_Entries_Can_Be_The_Same_Object()
+        {
+            var o = new object();
+
+            var builder = new NodeStateTable<object>.Builder();
+            builder.AddEntries(ImmutableArray.Create(o, o, o), EntryState.Added);
+            var table = builder.ToImmutableAndFree();
+
+            var expected = ImmutableArray.Create((o, EntryState.Added), (o, EntryState.Added), (o, EntryState.Added));
+            AssertTableEntries(table, expected);
+        }
+
+        [Fact]
+        public void Node_Builder_Can_Add_Entries_From_Previous_Table()
+        {
+            var builder = new NodeStateTable<int>.Builder();
+            builder.AddEntries(ImmutableArray.Create(1), EntryState.Added);
+            builder.AddEntries(ImmutableArray.Create(2, 3), EntryState.Cached);
+            builder.AddEntries(ImmutableArray.Create(4, 5), EntryState.Modified);
+            builder.AddEntries(ImmutableArray.Create(6), EntryState.Added);
+            var previousTable = builder.ToImmutableAndFree();
+
+            builder = new NodeStateTable<int>.Builder();
+            builder.AddEntries(ImmutableArray.Create(10, 11), EntryState.Added);
+            builder.AddEntriesFromPreviousTable(previousTable, EntryState.Cached); // ((2, EntryState.Cached), (3, EntryState.Cached))
+            builder.AddEntries(ImmutableArray.Create(20, 21, 22), EntryState.Modified);
+            builder.AddEntriesFromPreviousTable(previousTable, EntryState.Removed); //((6, EntryState.Removed))); 
+            var newTable = builder.ToImmutableAndFree();
+
+
+            var expected = ImmutableArray.Create((10, EntryState.Added), (11, EntryState.Added), (2, EntryState.Cached), (3, EntryState.Cached), (20, EntryState.Modified), (21, EntryState.Modified), (22, EntryState.Modified), (6, EntryState.Removed));
+            AssertTableEntries(newTable, expected);
+        }
+
+        [Fact]
+        public void Node_Table_Entries_Are_Cached_Or_Removed_When_Compacted()
+        {
+            var builder = new NodeStateTable<int>.Builder();
+            builder.AddEntries(ImmutableArray.Create(1, 2, 3), EntryState.Added);
+            builder.AddEntries(ImmutableArray.Create(4, 5, 6), EntryState.Removed);
+            builder.AddEntries(ImmutableArray.Create(7, 8, 9), EntryState.Added);
+            var table = builder.ToImmutableAndFree();
+
+            var expected = ImmutableArray.Create((1, EntryState.Added), (2, EntryState.Added), (3, EntryState.Added), (4, EntryState.Removed), (5, EntryState.Removed), (6, EntryState.Removed), (7, EntryState.Added), (8, EntryState.Added), (9, EntryState.Added));
+            AssertTableEntries(table, expected);
+
+            var compactedTable = (NodeStateTable<int>)table.Compact();
```

Comment:
>(NodeStateTable<int>)table.Compact [](http://example.com/codeflow?start=33&length=34)

Should `Compact()` return `NodeStateTable<T>`?

Reply:
Yeah, this is tricky, because `Compact()` is called from the `DriverStateTable`, which doesn't know the `T`. I ended up making a non-generic interface with `Compact()` that returns that interface so it could call it in a type agnostic manner; all the other use cases we know the `T` so can cast it back. 

Open to suggestions on better approaches though.

Reply:
Perhaps `NodeStateTable<T>` could implement the interface explicitly.
```csharp
public NodeStateTable<T> Compact()
{
    // ...
}

IStateTable IStateTable.Compact()
{
    return Compact();
}
```

Reply:
Oh, yeah thats neat. thanks